File: oceanfft/components/WaveCascadeParameters.cs
using System.Text.RegularExpressions;
using Godot;

[Tool]
public partial class WaveCascadeParameters : Resource {
    // ## Denotes the distance the cascade's tile should cover (in meters).
    public delegate void ScaleChanged();
    public ScaleChanged scaleChanged;

    [Export] Vector2 TileLength{
        set {
            tileLength = value;
            shouldGenerateSpectrum = true;
            scaleChanged?.Invoke();
        }
        get => tileLength;
    }
    Vector2 tileLength = new(50, 50);
    // # Note: This should be reduced as the number of cascades increases to avoid *too* much detail!
    [Export(PropertyHint.Range, "0,2")] float DisplacementScale {
        set {
            displacementScale = value;
            scaleChanged?.Invoke();
        }
        get => displacementScale;
    }
    float displacementScale = 1.0f;
    // : # Note: This should be reduced as the number of cascades increases to avoid *too* much detail!
    [Export(PropertyHint.Range, "0,2")] float NormalScale {
        set {
            normalScale = value;
            scaleChanged?.Invoke();
        }
        get => normalScale;
    }
    float normalScale = 1.0f;
    // ## Denotes the average wind speed above the water (in meters per second). Increasing makes waves steeper and more 'chaotic'.
    [Export] float WindSpeed {
        set {
            windSpeed = Mathf.Max(0.0001f, value);
            shouldGenerateSpectrum = true;
        }
        get => windSpeed;
    }
    float windSpeed = 20.0f;
    [Export(PropertyHint.Range, "-360,360")] float WindDirection {
        set {
            windDirection = value;
            shouldGenerateSpectrum = true;

        }
        get => windDirection;
    }
    float windDirection = 0f;
    // ## Denotes the distance from shoreline (in kilometers). Increasing makes waves steeper, but reduces their 'choppiness'.
    [Export] float FetchLength {
        set {
            fetchLength = Mathf.Max(0.0001f, value);
            shouldGenerateSpectrum = true;
        }
        get => fetchLength;
    }
    float fetchLength = 550.0f;
    [Export(PropertyHint.Range, "0,2")] float Swell {
        set {
            swell = value;
            shouldGenerateSpectrum = true;
        }
    }
    float swell = 0.8f;
    // ## Modifies how much wind and swell affect the direction of the waves.
    [Export(PropertyHint.Range, "0,1")] float Spread {
        set {
            spread = value;
            shouldGenerateSpectrum = true;
        }
        get => spread;
    }
        float spread = 0.2f;
    // ## Modifies the attenuation of high frequency waves.
    [Export(PropertyHint.Range, "0,1")] float Detail {
        set {
            detail = value;
            shouldGenerateSpectrum = true;
        }
    }
    float detail = 1.0f;
    // ## Modifies how steep a wave needs to be before foam can accumulate.
    [Export(PropertyHint.Range, "0,2")]  float Whitecap {
        set {
            whitecap = value;
            shouldGenerateSpectrum = true;
        }
    }
    float whitecap = 0.5f; // # Note: 'Wispier' foam can be created by increasing the 'foam_amount' and decreasing the 'whitecap' parameters.
    [Export(PropertyHint.Range, "0,10")] float FoamAmount {
        set {
            foamAmount = value;
            shouldGenerateSpectrum = true;
        }
    }
    float foamAmount = 5.0f;
    public Vector2I spectrumSeed = Vector2I.Zero;
    bool shouldGenerateSpectrum = true;
    public float time = 0f;
    float foamGrowRate;
    float foamDecayRate;
    // # References to wave cascade parameters (for imgui). The actual parameters won't
    // # reflect these values unless manually synced!
    // var _tile_length = [tile_length.x, tile_length.y]
    // var _displacement_scale = [displacement_scale]
    // var _normal_scale = [normal_scale]
    // var _wind_speed = [wind_speed]
    // var _wind_direction = [deg_to_rad(wind_direction)]
    // var _fetch_length = [fetch_length]
    // var _swell = [swell]
    // var _detail = [detail]
    // var _spread = [spread]
    // var _whitecap = [whitecap]
    // var _foam_amount = [foam_amount]


}

File: src/wave_cascade_parameters.rs
use godot::classes::class_macros::registry::signal;
use godot::prelude::*;
use godot::classes::Resource;
#[derive(GodotClass)]
#[class(base=Resource)]
pub struct WaveCascadeParameters {
    // So #[export] shows it to the editor
    // #[var] shows it to the code
    // #[func] shows a function to the code
    #[export]
    pub tile_length: Vector2,
    #[export(range = (0.0, 2.0))]
    pub displacement_scale: real,
    #[export(range = (0.0, 2.0))]
    pub normal_scale: real,
    #[export(range = (0.0001, 10.0, or_greater))]
    pub wind_speed: real,
    #[export(range = (-360.0, 360.0))]
    pub wind_direction: real,
    #[export(range = (0.0001, 1000.0, or_greater))]
    pub fetch_length: real,
    #[export(range = (0.0,2.0))]
    pub swell: real,
    #[export(range = (0.0,1.0))]
    pub spread: real,
    #[export(range = (0.0,1.0))]
    pub detail: real,
    #[export(range = (0.0,2.0))]
    pub whitecap: real,
    #[export(range = (0.0,10.0))]
    foam_amount: real,
    pub spectrum_seed: Vector2i,
    pub should_generate_spectrum: bool,
    pub time: real,
    pub foam_grow_rate: real,
    pub foam_decay_rate: real,
    base: Base<Resource>
}
#[godot_api]
impl IResource for WaveCascadeParameters {
    fn init(base: Base<Resource>) -> Self {
        godot_print!("Hello world!");
        Self {
            tile_length: Vector2::new(50.0, 50.0),
            displacement_scale: 1.0,
            normal_scale: 1.0,
            wind_speed: 20.0,
            wind_direction: 0.0,
            fetch_length: 550.0,
            swell: 0.8,
            spread: 0.2,
            detail: 1.0,
            whitecap: 0.5,
            foam_amount: 5.0,
            spectrum_seed: Vector2i { x: 0, y: 0 },
            time: 0.0,
            foam_decay_rate: 0.0,
            foam_grow_rate: 0.0,
            should_generate_spectrum: true,
            base
        }
    }
}
#[godot_api]
impl WaveCascadeParameters {
    #[signal]
    pub fn scale_changed();
}